### ISSUE TITLE
Filtering of Incoming messages

### ISSUE THREAD

**Issue by hitam4450** [nextcloud/mail#44]
for organizing message it would be very useful to incorporate a filtering mechanism, based on different conditions e.g. specific account, specific sender, Subject, etc., which will be applied to all incoming messages!

regards, hitam4450

Todo
- [x] Basic sieve support https://github.com/nextcloud/mail/pull/4472
- [ ] Editor with syntax highlighting
- [x] UI that can build rules

![badge](https://api.bountysource.com/badge/issue?issue_id=37588385)

**Comment by zeugmatis** [comment]
This is a nice idea but it is important to draw a clear distinction between the [MUA (Mail User Agent)](https://en.wikipedia.org/wiki/Email_client) and [MDA (Mail Delivery Agent - filtering)](https://en.wikipedia.org/wiki/Mail_delivery_agent).  MDA solutions on the server side already exist e.g. procmail, maildrop, Dovecot LDA module and others - if you have SSH access to your server, you can set up your rules to do this today.  😄  

Using an MDA, your filters are only maintained in one place and you have consistent behavior across _all_ of your devices / email clients.  Otherwise you have to maintain the same rulesets across all your devices, and can have confusion / possible race conditions when checking on e.g. the cellphone, then check on webmail and suddenly emails moved around.  Finally, when using an MDA rules will execute immediately when mail is delivered - so things like vacation autoresponders, forward-to based on From: etc. do not have to wait for you to login.  

It is a good idea, but warrants it's own app I think?  Since MDA's can have a lot of accidental complexity - it seems simple at first glance but can get so much more complex, see for example the number of options in the native clients like apple's Mail.app.

**Comment by zeugmatis** [comment]
BTW if this is to be pursued either in Mail app or a separate MDA app for NC, I would like to mention [Sieve, which is based on RFC 5228](http://sieve.info/) - already has existing support in many clients/servers/existing API's etc.

**Comment by laurentdinclaux** [comment]
Of course, a sieve UI is much needed.
A library that could be used: http://sieve-php.sourceforge.net/

**Comment by pierreozoux** [comment]
Also be able to filter read/unread on inbox.

**Comment by ediazcomellas** [comment]
A sieve management screen is pretty much a standard in webmail systems. I am willing to setup a bounty for this feature. Please contact me for details.

**Comment by jancborchardt** [comment]
@ediazcomellas you can donate to the bounty at https://www.bountysource.com/issues/37588385-filtering-of-incoming-messages :)

cc @hitam4450 @zeugmatis @laurentdinclaux @pierreozoux if you’d like this feature too, any donation or contribution is appreciated! :tada:

**Comment by ediazcomellas** [comment]
Is there anyone willing to make a try on this issue? I've expanded the bounty to $125. I hope it helps

**Comment by jancborchardt** [comment]
@ediazcomellas thanks a lot for the bounty! :)

@nextcloud/mail @ChristophWurst @Gomez @lcalaresu anyone up for solving this? :)

**Comment by ChristophWurst** [comment]
FYI: there's a pretty high bounty set for this ticket. Thus, a potential bounty hunter might be interested in implementing this feature. If so, please contact us first here in this ticket and let us know *how* you want to solve this problem. This will ensure that it's done right and you don't waste too much time on a solution that won't be integrated. Hope that makes sense. Thanks.

**Comment by ediazcomellas** [comment]
I've raised the bounty to $250... anyone willing to solve it? @denim2x @ChristophWurst  @Gomez @lcalaresu @nextcloud/mail

**Comment by MG2R** [comment]
Added another 50 to the bounty. Would love this feature.

**Comment by ediazcomellas** [comment]
I don't know why it is not an attractive bounty. Besides, there is a reference implementation in horde that can be used as a guide.

**Comment by jancborchardt** [comment]
We tweeted about the issue to get people involved: https://twitter.com/Nextclouders/status/941701115195940865 – please retweet and spread the word to help make it happen. :smiley:

**Comment by nado** [comment]
https://github.com/Rainloop/rainloop-webmail has support for sieve filters, its another app, but it may be worth checking how to improve integration of rainloop within nextcloud or get a similar sieve ui from them.

**Comment by muppeth** [comment]
Rainloop from what i saw has a specific way of saving and formatting filters. It would be good to follow the same scheme so that future migrations are painless.

On 15 December 2017 17:13:39 CET, Corentin Pazdera <notifications@github.com> wrote:
>https://github.com/Rainloop/rainloop-webmail has support for sieve
>filters, its another app, but it may be worth checking how to improve
>integration of rainloop within nextcloud or get a similar sieve ui from
>them.
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/nextcloud/mail/issues/44#issuecomment-352045425

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by pierlon** [comment]
Hey @nextcloud/mail, I'd like to implement this.

Should this be under Settings as a section, or should it have its own page? I plan to use the [pear/net_sieve](https://packagist.org/packages/pear/net_sieve) package btw.

**Comment by ChristophWurst** [comment]
^ @jancborchardt

**Comment by jancborchardt** [comment]
@pierlon cool stuff! :) Should definitely be in the Settings as a section. That is, the account-specific settings which can be opened through the 3-dot-menu, right? :)

**Comment by pierlon** [comment]
@jancborchardt Right 👍

**Comment by loxK** [comment]
@pierlon have a look at rainloop's implementation. I find their filter UI great and capable. 

I just forgot to save some new rules a couple of time making me loose them, but I guess it is better than a mistake leading to a "lost in the folders jungle" email because of that.

**Comment by pierlon** [comment]
@loxK OK thanks.

**Comment by kylepwarren** [comment]
Yeah this is pretty crucial. when I downloaded it, it was IMAP only. not sure if pop3 is supported now. either way, straight from host your own mail server - really needs filters. Probably hard to make. not sure $250 is enough. I'd do it for like $5000

**Comment by kylepwarren** [comment]
Would be cool if was built into cron or something, so this could be how you managed your IMAP mail (filters/sorting)

**Comment by kylepwarren** [comment]
To the person who said dovecot can do this already: Do you expect any email user to create a dovecot mail rules congif?

**Comment by loxK** [comment]
Nextcloud being about self-hosting I think most of us are using the mail app on self hosted mail servers. 

Sieve server-side filtering rules aren't exclusively supported by Dovecot, [see here](http://sieve.info/servers). Also Sieve server side filters are what most webmails add support for when talking about filtering (roundcube, rainloop, sogo etc.)

**Comment by kylepwarren** [comment]
oh. In that case. sieve rules.

**Comment by kylepwarren** [comment]
Was thinking about this again: sieve... not everyone will have access to that. I have to think eventually pop3 being supported would be a plan. Also, pretty sure a cron task wouldn't scale well. Feels funny, but I think the decision making should be offset to client - I guess it is just like every other email client. I guess it makes sense

**Comment by kylepwarren** [comment]
is there a js sieve library?

**Comment by loxK** [comment]
POP3 support is out of that issue scope. Nextcloud Mail is an IMAP client for now.

**Comment by ediazcomellas** [comment]
IMHO sieve is the way to go. Server-side filters are extremely useful: email gets delivered to the correct folder automatically. It doesn't matter if you open the mail with thuderbird, netxtcloud or your email client in your mobile phone.

**Comment by kylepwarren** [comment]
yeah, but not just server side, just in case they cant change their config - what if that server doesn't support sieve, or pretty sure not all sieve [server] support includes remote management by regular user. I saw a thunderbird plugin that looks like it does a good job - in the screen shots. Haven't tried it. See, I'm not sure if I'd have to install extra stuff for dovecot to support the remote management of sieve configs. I bet some people would have to. So, it should have client side fall back, feedback, instructions on how to edit the config, etc.

**Comment by ediazcomellas** [comment]
Server side and client side are two different tasks.  I will consider this task done, and concede the bounty, if the server side is done. 

Being nextcloud about selfhosting, there is a good chance that the admin can add sieve in the server, and any commercial service that is worth the money will already have sieve in the imap server anyway.

**Comment by ediazcomellas** [comment]
Anyone willing to take this issue? I really want to give the bounty :)

**Comment by TheLastTech** [comment]
Is this bounty still open?

**Comment by e-alfred** [comment]
@TheLastTech As there is no PR and this isn't closed, I would say yes.

**Comment by LemonAndroid** [comment]
Hey @e-alfred @kylepwarren,

Any progress on this and are you interested in teaming up for completing this?

cc @ChristophWurst

**Comment by PanderMusubi** [comment]
You can add this line at the top of the issue description, it will update itself

    ![badge](https://api.bountysource.com/badge/issue?issue_id=37588385)

and it looks like

![badge](https://api.bountysource.com/badge/issue?issue_id=37588385)

**Comment by LemonAndroid** [comment]
@PanderMusubi But it's a lie, the acutal bounty that can be cashed out is ~630$ because of the 10% bountysource fee.

**Comment by PanderMusubi** [comment]
🤔 Perhaps time for an API on top of this API that deducts the fee.

**Comment by LemonAndroid** [comment]
@ediazcomellas Why is this comment marked as outdated? :D https://gyazo.com/f9b55c46b4523bfefdc8c0b8904dd133

**Comment by ChristophWurst** [comment]
> You can add this line at the top of the issue description, it will update itself

The issue description should update automatically. I notified the person who set this up to check if the integration has to be-enabled.

**Comment by ediazcomellas** [comment]
@LemonAndroid I have no idea... The bounty is still open and awaiting for someone to take it. Please note that only a solution that is accepted upstream will be awarded the bounty.

**Comment by Krymson** [comment]
Hi - if no-one else is working on this, I'd like to volunteer to have a crack at it.

**Comment by ChristophWurst** [comment]
@Krymson awesome. Please let us know which approach you plan on implementing before starting your actual work so that we can make sure it's actually something we'd eventually integrate. Also, please see #1045 because we're restructuring/rewriting parts of this app at the moment and this might conflict.

**Comment by Krymson** [comment]
It will be a settings page to manage sieve filters on the mail server via the managesieve protocol described in RFC5804. 

Additionally there will be a page to set up a simple out of office notification using the same mechanism (but making it obvious as opposed to something you can do with filters).

Do you need more detail?

**Comment by LemonAndroid** [comment]
@ChristophWurst @Krymson Any progress on this?

**Comment by ChristophWurst** [comment]
> It will be a settings page to manage sieve filters on the mail server via the managesieve protocol described in RFC5804.

You might want to wait for the completion of #1045 in this regard.

Other than that I'd like to see how you'd technically implement this. How will you build the UI? How will you configure sieve rules on the IMAP server? How do you handle servers that don't support sieve filtering?

Thanks.

**Comment by nado** [comment]
So basically, still waiting? Rainloop for nextcloud is not supported in nc14 and official mail app doesnt support filters.
I guess I’ll wait on nc13 for some time this year.

**Comment by LemonAndroid** [comment]
@nado why is rainloop not supported in nc14 and why is nextcloud/mail superior to rainloop?

**Comment by nado** [comment]
@LemonAndroid rainloop is not officially maintained for nc, the current port is incompatible with the drop of deprecated API in nc14.

Rainloop support sieve filters, thats the only feature I've considered up til now to compare rainloop and nc/mail, because I rely heavily on them.

**Comment by nado** [comment]
Im not sure to understand what you say. If I have to, I'll switch back to a "rainloop only" instance.
But my hope was to have it more integrated in my current nc install (login session, file picking for attachment, less checks needed to keep it all updated, etc).
I'd be happy to switch to nc/mail when the filters are developed because thats basically all Im waiting for.

**Comment by LemonAndroid** [comment]
@nado I'm saying rainloop could be easily integrated to share login etc. with nc install.

**Comment by nado** [comment]
@LemonAndroid yes it *could*. Problem is manpower. So I'm looking for alternatives since I dont have time for personally managing it right now.
See https://github.com/pierre-alain-b/rainloop-nextcloud/pull/64#issuecomment-403088925

**Comment by LemonAndroid** [comment]
@nado thanks for linking the rainloop-nextcloud repo, let's continue discussion over there: https://github.com/pierre-alain-b/rainloop-nextcloud/issues/72

**Comment by loxK** [comment]
I have the same rainloop dilemma. Just learning it won't be supported in NC14... 😫

Every webmail app has sieve filters support. Sogo, rainloop, roundcube etc. It is the de facto standard for IMAP mail filtering.

**Comment by LemonAndroid** [comment]
@laurentdinclaux i was planning on putting up a bounty for the rainloop-nextcloud#45 issue for making it work on nc14. would you chime in? I thought 10$ each for now.

**Comment by plinss** [comment]
@LemonAndroid I'd be willing to put up $50 toward the bounty of getting rainloop working in NC14

**Comment by LemonAndroid** [comment]
@plinss I'll match that.

**Comment by ChristophWurst** [comment]
Please discuss these rainloop topics in the rainloop repo. Thanks :v:

**Comment by Sanookmakmak** [comment]
There are a lot of solutions already done, just install sieve:

http://sieve.info/clients

**Comment by Krymson** [comment]
@ChristophWurst - firstly, I don't plan on supporting servers that don't support managesieve. In the same way that Nextcloud doesn't support mail servers that don't support IMAP. 

From a UI point of view, I'm expecting that I'll provide ready to use rules and the parameters require for them rather than access the sieve language - in my mind this is an end-user tool, not poweruser/admins. For this reason I'll short-circuit the most common functions, such as setting an out-of-office message.

As for how technically - I need to investigate how the frontend works in Nextcloud. This would be my first Nextcloud contribution, and I need to familiarise.

**Comment by ChristophWurst** [comment]
> @ChristophWurst - firstly, I don't plan on supporting servers that don't support managesieve. In the same way that Nextcloud doesn't support mail servers that don't support IMAP.

We at least have to show a info that rules cannot be edited on these server and disable the rules dialogue.

**Comment by Krymson** [comment]
Sure, that's a given. I'll have to add some settings in the account settings dialog.

**Comment by ediazcomellas** [comment]
I am ok with a managesieve approach. Sieve is a sensible standard. 
Please check how roundcube manages several sieve scripts, allowing to activate one or another. 
![screenshot_20180905_190935](https://user-images.githubusercontent.com/3808234/45109252-4d231180-b13f-11e8-806f-bf26e1f99213.png)

**Comment by ytzelf** [comment]
@krymson are you still working on this?

**Comment by omocinteractive** [comment]
I installed Dovecot/Sieve and use Thunderbird to edit the sieve-filters for each account. I think it's a good way. It would be nice to have a editor in the nextcloud mailer, of course ....

**Comment by plinss** [comment]
I agree that filtering via Sieve on the mail server is the proper approach here. FWIW, I've been using RainLoop to edit sieve filters. Much better UX than Thunderbird, IMO.

Having a decent UI to edit Sieve filters via the managesieve protocol built in to the standard mail app would be great.

**Comment by omocinteractive** [comment]
This is also a reason i use Roundcube (was installed on my Plesk system and has a better UI for multiple accounts) ...

**Comment by ediazcomellas** [comment]
Hi, 

No, Sieve only acts on message receival for a single account, so it wont work on external accounts or work across several local accounts. 

Being mailserver-based, sieve is ideal if you access your email from several interfaces (which is the case of lots of users, using web, desktop and mobile MUAs).



> I raised bug #1101, but it was closed as duplicate of this bug. However, from reading above comments it seems that server side filtering is preferred through sieve. This makes me wonder two things:
> 
>     * Can sieve on the server side apply message filters across multiple accounts?
> 
>     * My use case is using mail app to read many external accounts, such as Gmail, outlook, and a few others, including private ones. AFAIK most public mail services doesn't support sieve server side. In this respect I think my original #1101 request should remain open.
> 
> 
> Also, I do not think that most users that self host also self host email. Just think about the press release about expanding nextcloud on consumer devices in Japan. Why limit such an important mail feature to advanced users.

**Comment by omocinteractive** [comment]
> * Can sieve on the server side apply message filters across multiple accounts?
yes, but you have to add the filtersetting for each account (copy and paste directly on the host or export/import in your editor). i forward mails from gmx, gmail to a self-hosted account with sieve filter.

**Comment by ytzelf** [comment]
I'm currently using roundcube and there is a well made and functionnal managesieve plugin here [https://github.com/roundcube/roundcubemail/tree/master/plugins/managesieve](url) . To alleviate above concerns about sieve complexity, the way it works is with a graphical user-friendly filter editor translating comboboxes in sieve syntax.
I'd be really interested in this as I want to switch to nextcloud for contacts and calendar purpose but can't live without client-side edition and creation of server-side filters. I'll add 20€ to the bounty.

**Comment by ytzelf** [comment]
> 
> 
> "I'd be really interested" - i too. I use a card-dav and web-dav in roundcube plugin to interact with nextcloud, but it would be better to do this in nextcloud mailer (if someday the signature feature available).

Then maybe add to the bounty? if it goes high enough someone will pick it up ^^

**Comment by omocinteractive** [comment]
... i still waiting for the confirmation mail (greylisting active).

**Comment by omocinteractive** [comment]
... done.

**Comment by daniilborovkov** [comment]
Hello. Is now someone except me working on this issue?

**Comment by TheLastTech** [comment]
Yeah, i don't do bounties after I did one and they refused to pay.

**Comment by LemonAndroid** [comment]
@TheLastTech @nextcloud/appstore refused to pay?

**Comment by LemonAndroid** [comment]
@daniilborovkov can i join you?

**Comment by ediazcomellas** [comment]
As one of the posters of the bounty, I feel a bit frustrated. After a loooong time, and with a quite high bounty after all, I don't see any progress on the issue. Is there anyone really working on it?

**Comment by LemonAndroid** [comment]
@ediazcomellas do you . plan to pull funding? is that possible?

**Comment by LemonAndroid** [comment]
I'd like to start working on this, but it's difficult alone.

**Comment by LemonAndroid** [comment]
At least we got the conversation sparking up again.

**Comment by ediazcomellas** [comment]
I really want to pay the bounty. I only put two conditions: work with sieve and produce a solution accepted upstream. I won't pull the bounty (it is not possible). I just feel a bit frustrated :)

**Comment by LemonAndroid** [comment]
I can understand.

**Comment by LorbusChris** [comment]
I added a little to the bounty. Somebody, please do this!

**Comment by LemonAndroid** [comment]
how about using ruby/truemail instead of sieve?

https://github.com/rubygarage/truemail

**Comment by loxK** [comment]
> how about using ruby/truemail instead of sieve?
> 
> https://github.com/rubygarage/truemail

How is this related?

**Comment by LemonAndroid** [comment]
@loxK isn't it much like sieve?

**Comment by loxK** [comment]
It is totally unrelated it is an email validator for the Ruby programming language...

**Comment by ediazcomellas** [comment]
Using sieve is an obvious choice: let the email server do the work. Email will be stored in an IMAP server anyway, so why not use it for filtering too? Most providers support sieve (all the serious ones) and it is very easy to setup one if you self host your email. All it takes is a parser/editor of sieve scripts and a way to deploy them (sieve has a protocol for that). Even better: you can borrow all the necessary the code from Horde, roundcube and many other webmail programs out there.

**Comment by LemonAndroid** [comment]
@loxK how is it totally unrelated? sieve is also a dsl. ruby is a great choice for dsls. isn't truemail doing email validation as well?

**Comment by LemonAndroid** [comment]
btw, i started doing the bounty.

don't worry, i'll use sieve.

Does any1 want to help/join?

**Comment by Gatak** [comment]
Great to see some discussions again. There are a few things sieve cant deal with. For example cross account filtering. Also I don't think that most major email providers like Gmail supports sieve.

**Comment by LemonAndroid** [comment]
http://eagle.botcompany.de progress can be seen here

**Comment by ediazcomellas** [comment]
> Great to see some discussions again. There are a few things sieve cant deal with. For example cross account filtering. Also I don't think that most major email providers like Gmail supports sieve.

Most self-hosters (and Nextcloud is all about self hosting) can implement it.  Gmail, like all the big guys, don't play well with others.  Cross account filtering is something that requires explicit support in the client, and it sounds just too complicated for me. I can live with a simple sieve support (and, btw, it is one of my two requirements to release the bounty).

**Comment by Gatak** [comment]
> > Great to see some discussions again. There are a few things sieve cant deal with. For example cross account filtering. Also I don't think that most major email providers like Gmail supports sieve.
> 
> Most self-hosters (and Nextcloud is all about self hosting) can implement it. Gmail, like all the big guys, don't play well with others. Cross account filtering is something that requires explicit support in the client, and it sounds just too complicated for me. I can live with a simple sieve support (and, btw, it is one of my two requirements to release the bounty).

Yes, you're right. Understandably difficult too. Could one way of seeing this be that Nextcloud is the client that uses IMAP to download from various accounts (like Gmail, Hotmail..) and applies the sieve locally? A sieve client would talk to NC and not the third party IMAP servers? It's just a thought. Thanks.

**Comment by morph027** [comment]
Right now it's just a client which does not download. One could "easily" setup a self hosted server including sieve which fetches from all remote accounts to apply cross account rules.

**Comment by ediazcomellas** [comment]
@morph027 is right. The current mail application in Nextcloud is just a webmail that talks IMAP to the server. What is needed here is a "settings" section that allows to edit/deploy sieve scripts to the email server. Let the server do the heavy work. Creating a new implementation of sieve in the client can be a nightmare, and I doubt it would be useful. 

Filtering in the server (sieve) makes all views of your email (either using a desktop client, any webmail client or a mobile phone app) consistent and asynchronous (filtering happens on mail delivery, not when you open the mailbox with your email app). That is why I find it is the best place to do the work. 

Besides, letting the server do the filtering makes this request much easier to accomplish.

**Comment by ChristophWurst** [comment]
@ediazcomellas yes, thanks for the summary. This is exactly what we should do there and also the only solution I'd allow integration into our code base :wink: 

Reminder to whoever wants to work on this: let us know what and how you want to implement it. Thanks.

If nobody is interested I can have a look at this feature after summer vacation in a few weeks. As I'm a Nextcloud employee the Bounty would be put back onto other tickets as per usual policy.

I'm closing resolved discussion comments now to keep the thread clean for future readers.

**Comment by LemonAndroid** [comment]
@ChristophWurst I started working on the issue.

**Comment by LemonAndroid** [comment]
But now i'm getting a bit sceptical. A few comments ago I already mentioned I started working on this issue + another commenter raised concerns about Nextcloud paying out bounties + it seems the preferred way for bounties is to be "collected" by employees and put back into the pool.

You kinda completely defeated the purpose of bounties by this logic.

**Comment by LemonAndroid** [comment]
I also wish to not be a member anymore, I don't want to be associated with Nextcloud at this point.

**Comment by stefan-reich** [comment]
@ChristophWurst: @LemonAndroid posted here that he works on the issue and you ignored it. Is that the way you treat collaborators?

**Comment by kesselb** [comment]
Please calm down :sunglasses: 

> A few comments ago I already mentioned I started working on this issue + another commenter raised concerns about Nextcloud paying out bounties + it seems the preferred way for bounties is to be "collected" by employees and put back into the pool.

1) I have the impression that the person who wrote that statement never made a bounty for nextcloud. Not sure why he is even here. You may check his github profile. 

2) I made a ticket with bounty once (https://github.com/nextcloud/mail/pull/1061) and got the bounty.

**Comment by morph027** [comment]
I guess not by purpose, probably by accident. Also being member of the nextcloud github organisation does not make you an employee. I'm pretty sure you'll get your bounty once it's done.

**Comment by LemonAndroid** [comment]
@morph027 @kesselb thanks for calming me down.

**Comment by loxK** [comment]
@LemonAndroid  please read what @ChristophWurst said. He said to get in touch to anybody taking the work. **If nobody does**, he **may** do it after summer vacation. It is pretty clear and nothing is wrong with that.

If you want to do it, well just **do it**, this is how open source works.

![image](https://user-images.githubusercontent.com/136687/59918781-7bc16d00-9471-11e9-919c-42688d1938e8.png)

Now please stop useless comments that notify everybody in the thread.

**Comment by LemonAndroid** [comment]
@kesselb I also did a small bounty like yours tho: https://github.com/nextcloud/deck/issues/259.

I'm sceptical about the payout of big bounties.

**Comment by LemonAndroid** [comment]
@loxK you're hurting the discussion with wanting to throttle it down. you can stop subscribing to notifications. easily.

**Comment by LemonAndroid** [comment]
This is it with you people. you want users and participators, but you're the ones at fault when everything get's stale because discussion was not possible.

And yes you fools, I know there is an edit button.

**Comment by ChristophWurst** [comment]
Sorry for my bad communication here. Of course I've read @LemonAndroid. Unfortunately so many people have tried to claim bounties before they even had a look at the code. You can search the issue tracker for bounty tickets and you'll find many where people said they were working on somehting, others held back and in the end there was not a single line of code.

My statement was only regarding the case that nobody would actually pick it up.

@LemonAndroid as soon as you have some code please open a PR so we can discuss the details there.

> But now i'm getting a bit sceptical. A few comments ago I already mentioned I started working on this issue + another commenter raised concerns about Nextcloud paying out bounties + it seems the preferred way for bounties is to be "collected" by employees and put back into the pool.

That is not true. As @morph027 pointed out this is only done when an employee implements a feature which a bounty has been set for. Naturally this has happened many times because when I post bounties then I put them on tickets that make most sense and not some random edge case feature. Hence I, as the main developer, occasionally do stuff which others requested and offered their monetary support for.

I've only rejected paying bounties when a contributor provided an incomplete solution, usually claimed before it was even integrated. But those conflicts were resolved quickly and AFAIK all bounties have been paid out.

IMO the purpose of bounties was never to let them grow as much as it has here, but so be it. When I distribute money I do in batches of small amounts so it's not concentrated on a few tickets but allows earning a few bucks on as many tickets as possible: https://twitter.com/ChristophWurst/status/1120707552856215553.

:)

**Comment by LemonAndroid** [comment]
10 usd minus bountysource fees was never earning.

I still like your professionally-worded response.

**Comment by ChristophWurst** [comment]
I'm sorry to hear that. I was not aware of the fee as I never had a bounty paid out. But I think with the $800 here the $10 are negligible :see_no_evil: In the long run we should move away from Bountysource. We've had too many issues with it recently (both technically and similar to what you raised).

Now let's stop the off-topic discussion and focus on making Mail an awesome e-mail app :rocket:

**Comment by LemonAndroid** [comment]
just one last "off-topic": Bountysource charges a 10% fee on all withdrawals from your Bountysource account.

**Comment by pierlon** [comment]
@LemonAndroid started anything?

**Comment by LemonAndroid** [comment]
@pierlon I tried and yet I have issues setting up the `mail` app on the server. Some problem with `npm`/`nodejs` and probably the old version of Ubuntu, `16.04`. Once I figure that out, I'll continue with a fork of the solution you started.

**Comment by mmittelb** [comment]
@LemonAndroid any progress?

**Comment by LemonAndroid** [comment]
Nope

**Comment by mmittelb** [comment]
If there is no progress on @ghost side, I'd like to take over the issue. I'm starting tomorrow. 
Cheers

**Comment by kesselb** [comment]
@mmittelb do it :+1: The user `@LemonAndroid` deleted himself from github. To keep issues and comments such accounts are renamed (or merged) to `@ghost`.

**Comment by mmittelb** [comment]
@kesselb oh good to know 
@plinss 
To 1:  STARTTLS option is planned. 
To 2: I only know the roundcube sieve client. But I'll have a look at rainloop before implementing the UI.

**Comment by ezii123** [comment]
Thanks for taking over this, I'm the @ghost btw 🗡

**Comment by loxK** [comment]
Roundcube does a pretty job as it gives access to all sieve rules. Rainloop only manages sieve rules in its own filters file:

![image](https://user-images.githubusercontent.com/136687/64058808-7fedd180-cbfc-11e9-9b26-e085225760c4.png)

**Comment by mmittelb** [comment]
Little update:
I set up the server with the mail app and started to experiment with the frontend. Since I'm new to developing webapps this may take some time. 
Frontend:
I figured out how to add an option to the settings menu (AppSettingsMenu) and will start to work on the view. The idea is, to have a dropdown list to select the account. Once selected there will be a roundcube-like gui (I just like it better) to edit filters and an expandable settings menu to specify server, port and STARTTLS options plus testing button. Default should be the imap server, port 4190 and STARTTLS enabled with cert checking. As soon I gat something working I'll push it to git. Should I branch the current master or how are you using git?
Backend: 
Didn't do much yet. But I will probably use the horde-managesieve library.

**Comment by loxK** [comment]
I had to have a deeper look in Sieve rules and webmails for another purpose and have some thoughts.

In terms of "Sieve", activating a filter set makes a symlink from sieve-script -> filtersetName.user.sieve. 

Rainloop (and maybe other clients) maintains its rules in its own filter set file with some hash ids. What this implies is that any modification done from another client may make it unreadable by Rainloop afterwards.

So my idea would be that the GUI will present  filter sets in a similar in what Roundcube does. But:
1. the  filter sets GUI would support drag and drop sorting
2. one "main" filter set would be hidden, its purpose would be to handle other filter sets activation and sorting using [url="https://wiki2.dovecot.org/Pigeonhole/Sieve/Examples#Include_scripts"](Sieve's include)

**Comment by ediazcomellas** [comment]
We use roundcube extensively here, so if what you implement supports it, It would be perfect (for us).

**Comment by ezii123** [comment]
Will the rules be saved to the NC database?

**Comment by plinss** [comment]
Hopefully the rules will only live in the sieve files, so they don't get out of sync if edited by another client.

Another ask would be that the sieve server settings (and possibly the filters?) be settable via occ (I'm not sure how much of that you get for free).

**Comment by loxK** [comment]
@ezii123 Sieve rules in the NC database is not a good idea nor needed. The rules are stored on the mail server.

@plinss occ support seems rather complicated to me. You'll need to specify the user's email account you want to set filters rules for, the rules set file etc. But you can do global sieve filters or default filters and so on [on you mail server directly](https://wiki.dovecot.org/Pigeonhole/Sieve/Configuration#Executing_Multiple_Scripts_Sequentially) and script that easily.

**Comment by plinss** [comment]
@loxK at the very least we should be able to set default server parameters for managesieve as we do for the smtp and imap servers in config.php 'app.mail.accounts.default'. I don't feel strongly if that's in config.php or settable via occ. The general user won't/shouldn't know how to configure their sieve server.

**Comment by ezii123** [comment]
When will we see the first Pull Request?

**Comment by kesselb** [comment]
@ezii123 Have you read https://github.com/nextcloud/mail/issues/44#issuecomment-527294986?

@mmittelb There is a draft with some backend changes. Might be handy: https://github.com/nextcloud/mail/pull/1822

**Comment by mmittelb** [comment]
@ezii123 I'm not planing on saving the script on the server. I don't think there is any need for that. You will see the pull request as soon there is something to pull. It's taking some time setting up everything and getting to know the frameworks and concepts used. Please be patient.
@loxK Thanks for your ideas. I like both of them. But side-by-side use with Rainloop won't work. 
@plinss Hmm I'm not sure about the occ feature. I don't know yet, how hard it will be to implement and I don't really see a scenario in that this is needed. Do you have one?
@kesselb Thx I'll have a look.

**Comment by plinss** [comment]
@mmittelb the requirement for this is two-fold: 

1) we can already setup default parameters for smtp and imap servers, so that in an environment where all Nextcloud users use the same email server (and the mail server uses the same credentials), users can simply click the mail icon and see their email without doing any configuration on their end. We should be able to set the defaults for the sieve server so users can then simply use the filtering UI without any configuration as well.

2) most of Nextcloud can be setup automatically using tools like ansible by either specifying information in config.php or running occ to configure apps. The default sieve server information  should be able to be setup automatically as well. I really don't care if it's done via config.phg (like the default mail server info already is), or via occ. Given that pretty much all apps can be configured via occ I'm guessing you'll get this for free so long as you use a standard way of storing app config information.

I run a business where we setup and maintain Nextcloud instances for customers, and everything is done via automation. If this has to be manually configured by each (non-technical) user or by using the admin web UI it would be a non-starter for us to adopt.

**Comment by kesselb** [comment]
> we can already setup default parameters for smtp and imap servers

Sounds reasonable to me. Needs some getters and setters for the default values [here](https://github.com/nextcloud/mail/blob/master/lib/Service/DefaultAccount/Config.php)  (config.php to object)  and [here](https://github.com/nextcloud/mail/blob/master/lib/Service/DefaultAccount/Manager.php) (default config to account config). Given that this ticket is heavy I would not demand it. @ChristophWurst @ediazcomellas opinions (you're the top backers).

> via occ I'm guessing you'll get this for free so long as you use a standard way of storing app config information.

Yes. Just use `IConfig` and you are good.

**Comment by loxK** [comment]
> @loxK Thanks for your ideas. I like both of them. But side-by-side use with Rainloop won't work.

Roundcube may have issues too. No sure how it will show a file filed with includes.
In the current situation Rainloop and Roundube aren't interoperable either. Roundcube would break a saved Rainloop filters set on save and fails to parse conditions with brackets.

But, the more I think about it the more I like the idea of having a main file that has a list of user sorted and activated fiters set. The idea is so good to me that I think that, once implemented this way in Nextcloud as a proof of concept, pull requests to Rainloop and Roundcube to make them comptatible with it has good chances to make its way to their code base.

**Comment by mmittelb** [comment]
@plinss On scriptable server information I agree and that's easy to implement. I thought you mean the sieve script handling, which is in my opinion easier to do directly communicating with the sieve server.
I think it's best to do the script rendering in the browser to make the page more responsive. Any opinion on that?

**Comment by plinss** [comment]
I did mention possibly setting some initial filters, but that was just a random side thought as I was typing. I agree if there's a need for that we can always just set some initial filters directly on the sieve server. My main concern was the configuration of the server settings.

As far as doing the script rendering client side, I agree, that'll be a much more responsive experience for the user.

**Comment by ediazcomellas** [comment]
Initial filters are a job of the server. Don't waste efforts on that.

**Comment by mmittelb** [comment]
@pierlon I'll fork your repo and merge.
Is the filters ressource fully working?
Is there a way to check sieve files by the mailserver? I think managesieve does have such functionality.
My view is integrated into the settings menu and I think it's better to leave it that way. For the server configuration, I like your solution better though.

**Comment by plinss** [comment]
Looking good so far.

One valuable action option will be to delete the message rather than move to trash. Commonly used for certain kinds of spam that there's no reason to keep in the trash.

A UI note, IMO the ability to edit the raw script is more of a power user feature that most users won't use (or understand), it might be better if that control was less prominent.

**Comment by ChristophWurst** [comment]
Is this jquery UI?

**Comment by jancborchardt** [comment]
Heya @mmittelb, thanks for your contribution! :)

So this seems to be very complex and complicated. With Nextcloud we aim to design stuff simply so regular people can use it. Editing the rules in plaintext format is not really accessible in that way.

What we need to do is start with the basics, like checking if the subject or sender matches or contains a specific text. Interface-wise this can be done very simply, similar to the Workflow redesign at https://github.com/nextcloud/server/pull/17506#issuecomment-542341901
And the "Filters" section can just be a section in the account settings.

From there on we can see what else is needed, but keep in mind all this needs to be maintained, so should be starting simple, with a minimum viable product.

What do you think? Also, it would be nice if you can open a pull request for this so we can use that to collaborate and discuss. :) Thanks!

**Comment by mmittelb** [comment]
Thanks for the input.
> delete the message rather than move to trash
I think the command is discard. I will implement it.
> raw script is more of a power user feature
Agreed. I implemented it as a fallback, because I don't want to add to many options to the parser. Imo too many options are confusing to the user. But I will think of a way to better disguise that button.
> Is this jquery UI?
No the dragging is vue-draggable which is a Vue implementation of sortablejs (https://github.com/SortableJS/Vue.Draggable)
> very complex and complicated
It's kind of a copy of the roundcube filter GUI. I always liked that one because it's straightforward for inexperienced user but at the same time powerful enough for the power users. I put the raw edit in the gif more to demonstrate the parser. The normal user shouldn't need to use that.
> checking if the subject or sender matches or contains a specific text
That is implemented.
> similar to the Workflow redesign at nextcloud/server#17506 (comment)
Could you be more concrete. I already tried to hide away most of the Sieve possiblities. To make it easier to make filters withoutuch knowledge.
> And the "Filters" section can just be a section in the account settings.
@pierlon suggested the same. Two reasons why I don't like it:
- Filters are just no server configuration option to me. I think it's cleaner to have its own section.
- The current UI would break the server configuration options design. But to implement the features we discussed in this thread I couldn't think of anything that would fit there. I didn't want to reinvent the wheel that's why I followed the roundcube way.
> keep in mind all this needs to be maintained
I understand. I tried to code it modular (using Vue components) and separated the parser from the page. I should definitely implement some tests. 
 > open a pull request
I used @pierlon work on the backend (https://github.com/nextcloud/mail/pull/1822) as a base. That's why I opened a pull request into his branch. Should I have my own pull request?

**Comment by jancborchardt** [comment]
@mmittelb ok, then we really have to check it. :)
Generally I'd say better look at Apple Mail or Google Mail rather than Roundcube for inspiration, as that is more widely used and tested.

> Filters are just no server configuration option to me

That's why it's called "Account settings" and has Signature and Edit mode too. :) See https://github.com/nextcloud/mail/pull/2119
Filters belong there.


Regarding the last point on branching, @ChristophWurst how should it be done?

**Comment by ChristophWurst** [comment]
> Regarding the last point on branching, @ChristophWurst how should it be done?

It should be pushed to *some* branch. I don't really mind which PR it is :)

**Comment by luvis** [comment]
I'm pretty new to navigating around github so I might just be lost but I can't seem to find any progress on this issue since june 2019, or is the progress going on somewhere else? I'm just curious because this is a super big deal if it gets implemented. It would make the email app so much more useful!

**Comment by Gunnar-GO** [comment]
anything new ?

**Comment by ChristophWurst** [comment]
Hey there,

I'd just like to let you know that due to the [ongoing issue with Bountysource](https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943) we're discussing ways to move to a different platform. As it looks right now, Bountysource might claim some of the money that was contributed by the community. We will try our best to prevent that. Historically we – the Nextcloud GmbH employees – simply put the claimed bounties back to other tickets until one was solved by a community member. So the money was always *from the community* and *for the community*. We'll therefore move to a better platform and currently don't post any new bounties until this situation is resolved. The current goal is to move/donate the money to [Nextcloud include](https://nextcloud.com/include/) but details will follow in the [forum post](https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943).

Hope this makes sense. If you have any questions please ask them on the forum. This comment will be copy-pasted to all open issues with bounties.

Christoph and the Mail team

**Comment by ediazcomellas** [comment]
> Hey there,
> 
> I'd just like to let you know that due to the [ongoing issue with Bountysource](https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943) we're discussing ways to move to a different platform. 

Please @ChristophWurst , claim the bounty and move it to somewhere useful. Bountysource has been a bad place from day one. I would rather see my money going to somewhere else in the nextcloud comunity.

**Comment by MG2R** [comment]
Please, claim the bounty, take whatever money you can, switch platforms and put the bounty up there. I posted my bounty end of 2017. No way that money is going to do anything useful if you don’t claim it. BS will drag this out until the deadline and then claim all the money.

**Comment by ChristophWurst** [comment]
The problem is that if not all 9 backers accept the bounty will remain on bountysource's side and we'll lose the money.

I've tried the *quick claim* on my own bounties like at https://github.com/nextcloud/mail/issues/317 but ironically the bountysource page is too broken to update the open/close state and I can't even claim the resolved issue at https://www.bountysource.com/issues/42324763-improve-caching-preloading-for-better-performance. It's really a mess.

**Comment by rubo77** [comment]
I had a chat on bountysource at freenode IRC a year ago, they helped me out and refunded my bounty by then

**Comment by MG2R** [comment]
Their site is a mess since the announcement. People must be waltzing in like herds.

@ChristophWurst what do you suggest I/we do? The money I pledged should go to Nextcloud. Preferably towards this issue but definitely towards your organisation, that’s why I said claim it. If that’s not possible, what’s the best way of helping out?

**Comment by HLFH** [comment]
ownCloud/Nextcloud since 2010 is the leading project in protecting personal data in the cloud era.
It has built a great community.

What I suggest is to:

 - instead of being manual in the process,  count how much backed money is remaining in BS ;
 - have a legal agreement with Bountysource Inc ;
 - in this agreement, make them agree to move all the remaining backed money to Nextcloud Include ;
 - wait for the transfer ;
 - be transparent as always ;
 - with Nextcloud Include, give backed money to contributors once they have pushed the required PRs.

1. Get all the backed money from BountySource and put it in Nextcloud Include.
2. Develop and internalize progressively a bounty solution for Nextcloud contributors.

And huge thanks to @ChristophWurst for the great rework of Nextcloud Mail.

**Comment by kesselb** [comment]
> Besides forum post, see main GitHub discussion about BS here: #44 (comment).

@HLFH no. Any discussion should be at https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943. The bountysource issue is completely unrelated to "Filtering of Incoming messages".

**Comment by dehnhardt** [comment]
To keep the hope alive, here is an illustrated intermediate result ;-) I curse Javascript and am still astonished what all this does...

![image](https://user-images.githubusercontent.com/5556832/85044268-aaaa1600-b18d-11ea-91a9-b473877b2f6b.png)

![image](https://user-images.githubusercontent.com/5556832/85044685-43d92c80-b18e-11ea-92e5-3b82484a21a1.png)

Holger

**Comment by dumblob** [comment]
For those needing filtering/Sieve **now**, there is an opensource solution [Sieve Editor]( https://github.com/thsmi/sieve/ ).

I'd also guess Sieve/filetring implementation in NC could also leverage the mature code base of Sieve Editor. What do you think @thsmi ?

**Comment by alphashuro** [comment]
Is the bounty on this still open?

**Comment by ChristophWurst** [comment]
https://www.bountysource.com/issues/37588385-filtering-of-incoming-messages is still open but please don't add anything to it. We're trying to move away from the bountysource platform: https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943

**Comment by jancborchardt** [comment]
@dehnhardt (and also maybe @ChristophWurst or @GretaD or @brueckner) do you want to sit together on figuring filtering out properly at some point?

We should start by collecting examples of how other systems do it, especially simple and easy to understand ones. Great examples are Sogo and Gmail:
- https://www.laptopmag.com/articles/filter-mail-gmail
- https://www.midgard.co.uk/how-tos/how-to-set-up-filters-for-email-in-sogo/

Additionally, Google does it nicely with different entry points to filtering like
- You can create filters from searches
- You can "Filter mails like this" to e.g. create a filter rule based on a sender, from a mail you are currently in

**Comment by plinss** [comment]
@jancborchardt Two other examples are [RoundCube](https://roundcube.net/) and [RainLoop](https://www.rainloop.net/). RainLoop also has a Nextcloud app so that would be a MVP target to get people to switch from it to the Mail app. Many people, such as myself, use it over the built-in Mail app for the filter support (and its superior feature set, and better UI, but Mail is catching up).

To me, one of the fundamental problems with sieve filters is interoperability. When the filters are written to the server, they're script, and few tools (if any) are able to parse the script back into the filter primitives.

Most tools, like RainLoop and RoundCube, also store filter metadata as comments in the script. That way, when they read the filter later, they can populate the UI and regenerate the same script. Unfortunately, no two tools I've ever seen agree on a format for that metadata. So if you create filters with one, then edit with another, many of your filter rules are lost (or at best don't show in the UI).

A huge win here would be standardizing on a single metadata format, and what script that metadata generates (and have open documentation for that format). Then get as many existing tools as possible to use that same format (submitting patches as necessary). Ideally work with authors of tools that already support sieve filters on the metadata format.

**Comment by loxK** [comment]
RainLoop has a compatibility issue. It stores hashes (as comments) in its sieve file. As such, editing this file with another client prevents editing it back in RainLoop.

What would be awesome is that the Nextcloud's implementation uses a master file that [includes](https://tools.ietf.org/html/rfc6609) a `nextcloud.sieve` file. If so, we could  then ask the RainLoop developer to modify its code in order to look for the master sieve file and roll its own [include](https://tools.ietf.org/html/rfc6609).

**Comment by ChristophWurst** [comment]
While I support the idea of making things as interoperable as possible searching for the perfect solution as the initial implementation will just make things more complex and push it down the roadmap. In my opinion it would be better if we first have something that works with our app and in another step we make it as interoperable as possible. If the fully interoperable solution was so easy, other solutions would have it …

**Comment by ediazcomellas** [comment]
I also support the interoperability idea, but wont want to wait one more year to get it running. OTOH, starting a parallel work to create a definition of the standard would be just great. I guess this is something to discuss in another thread. I would gladly contribute to such a working group.

**Comment by dumblob** [comment]
Disclaimer: I don't know Sieve very well.

How about not having a real **standard**, but rather a "standardized" (i.e. detectable) **convention** for a "main" Sieve file which would by any means **not** contain anything else than includes of other Sieve files ordered lexicographically (pretty much how Linux distributions started managing `/etc/` for different daemons during last 15 years).

The point would be, that if the included file can't be parsed by the given MUA, then that MUA has at least the opportunity to "get ahead" of other-MUA-filters and use the other-MUA-filters conditionally or even stop before other-MUA-filters.

This has the advantage, that if later there would actually be some standard, it'd stay compatible and useful as MVP for many use cases. On the other hand, despite being a small change this still requires all MUAs to make at least some minor changes to their behavior. But maybe Sieve projects like Pigeonhole could partially alleviate that by having defaults and recommendations to use such convention?

Is there any good place to discuss such potential standard/convention to not pollute this thread?

**Comment by plinss** [comment]
I accept that creating an open standard method of storing filter metadata takes time, and I agree that this feature shouldn't be gated on that. But at the least, I'd like to see some movement in that direction and a commitment from whoever implements this to help make it interoperable in the future. In fact, if an effort starts to create such a standard, you might find more people willing to help out.

Some suggestions:

1) Please do not simply create yet another non-interoperable format (unless you've studied all the existing formats and found them all wanting). This would just make the problem worse. At least start by following one existing format (ideally one that's easily detectable and extensible).

2) Ideally start with an exiting format that has open source code and refactor their metadata handling into a stand-alone reference library, that both they and this project can use. This could be done along with generating the documentation of the format and be used as a starting point for the standard. That way other tools can re-use the reference library to get interoperability instead of having to re-implement from scratch. (Without having looked closely at the actual metadata formats, I suggest starting with RainLoop simply because there are existing Nextcloud users.)

3) Given a common library, over time it can gain the ability to read other tools' proprietary metadata formats and convert into the standardized output format. Further improving interoperability without data loss.

4) Creating a standardized format doesn't have to be a formal standard (but it would be nice). Anyone can start such an effort by creating a GitHub repo and getting others to participate (I agree this is not the place for that effort). Ideally, invite authors of tools that have existing sieve filter handling and the original authors of sieve scripting. As it matures, you can search for a home for a more formal standard (I suggest IETF since that's where sieve lives).

I have some relevant skills and would be happy to participate in developing such a standard as time allows. Unfortunately I don't have the bandwidth to lead such an effort at this point in time, we just need one person to step-up and start it off...

**Comment by ChristophWurst** [comment]
Hey everyone :wave: 

we (Nextcloud GmbH) would like to finish this feature as part of our current [roadmap](https://github.com/nextcloud/mail/projects/10). @kesselb will help with that. Since this is a rather complex change we want to do it in two or three smaller chunks. First one would be the back-end support, acceptance criteria is passing unit and integration tests. In a second step we'll add a basic UI. Anything more advanced will be added in a third step.

Now I'd like to ask the original authors of the two open but stale PRs if it's ok if we build on their previous work. We won't take any bounty, as per usual, so the money will remain on the platform and I would like to ask those authors to later claim the remaining money when the PR for the second part is merged.

Is this ok for everone? I hope this way we can give credit where credit is due (or money) and from a bigger perspective we'll hopefully soon have our first release of Mail with this feature.

Please let me know if there are any objections or concerns.

Cheers

**Comment by ediazcomellas** [comment]
Hi!

It is ok for me. I would love to see this implemented. 

Best regards

**Comment by Samonitari** [comment]
> Hey everyone 
> 
> we (Nextcloud GmbH) would like to finish this feature as part of our current [roadmap](https://github.com/nextcloud/mail/projects/10). @kesselb will help with that. 
> ...
> Now I'd like to ask the original authors of the two open but stale PRs if it's ok if we build on their previous work.
> ... I would like to ask those authors to later claim the remaining money when the PR for the second part is merged.
> ...
> Cheers

Wow, this made me decide to finally donate to Nextcloud. This is quite mood-lifting attitude, especially after the onlyoffice "fiasco" if you can call it that.

**Comment by ChristophWurst** [comment]
> Wow, this made me decide to finally donate to Nextcloud.

Just make sure you're not being scammed because we don't take donations :wink:

**Comment by Samonitari** [comment]
> Just make sure you're not being scammed because we don't take donations

I can not decide between this being awesome, sad, or silly.
As I have no time for code contribution (beside not being versed in PHP), this would be the only way I could give back something, other than growing your reputation a little.

But thanks anyway for your words of caution XD

**Comment by dehnhardt** [comment]
Sorry I haven't commented in any form in the last few weeks - or was it even months. I have started a new job and that has kept me extremely busy.
I think it's great when the project is moved forward - be it with parts of my code or as a completely new development.
I think I will have some time again to participate - be it with stupid sayings or code ;-)

**Comment by ChristophWurst** [comment]
Update: we merged https://github.com/nextcloud/mail/pull/4472 today. It brings all the basics of Sieve filtering including the enhanced account settings, provisioning for organizations and a basic editor of the currently active script.

Based on this platform we can now think about an enhanced UI that can build rules. Hence I'll keep this ticket open for that remaining part.

**Comment by morph027** [comment]
Just tested successfully, thank you! :tada: 

```
root@mx:~# ls -la /var/mail/morph/
total 39
drwx--S---  4 vmail vmail   6 Mar  4 14:44 .
drwxrwsr-x  6 vmail vmail  10 Apr  7  2020 ..
lrwxrwxrwx  1 vmail vmail  21 Mar  4 14:44 .dovecot.sieve -> sieve/nextcloud.sieve
-rw-------  1 vmail vmail 258 Mar  4 14:41 .dovecot.svbin
drwx--S--- 12 vmail vmail  23 Mar  4 14:43 Maildir
drwx--S---  3 vmail vmail   4 Mar  4 14:44 sieve
```

```
Mar  4 14:41:17 mx dovecot: lda(morph)<28294><wM7CKH3jQGCGbgAAVh197Q>: sieve: msgid=<d8ea7b98-b51e-ff4f-dc36-79ab345de0f7@gmx.com>: stored mail into mailbox 'TEST'
```

**Comment by LorbusChris** [comment]
I think this should be closed out now, maybe a new issue could be opened for UI improvements.

@kesselb please feel free to claim https://www.bountysource.com/issues/37588385-filtering-of-incoming-messages

**Comment by Luke1982** [comment]
Roundcube has an excellent graphical implementation. Maybe we can borrow from that?

**Comment by sazanof** [comment]
Hello, everyone. I'm currently working on it.
Try parse sieve source code at frontend (but think that better to do in in backend) doesn't matter now, rexexp find me everywhere:)

I'll try to make easiest UI, that build easiest rules, and then I'll draft a PR.

**Comment by ChristophWurst** [comment]
That is awesome @sazanof!

I'm still not sure if we want/can "just" parse the sieve script. It's certainly too complex for a regex but needs a real lexer and parser. And then there are possibly a few syntax ways to defined the same semantics.

Therefore we want to first build a solution where a user can use Nextcloud Mail to define a rule. This will write a sieve rule in the typical sense, plus some metadata either as a comment our in our storage. With that metadata we will be able to reconstruct the UI state when the user re-opens the filter settings.
This is also what we have seen with another solution. That metadata could be some simple json encoded into a sieve comment.

It will always be either the UI filters or fully manual sieve scripting. The in between solution is too complex for a first version.

**Comment by sazanof** [comment]
@ChristophWurst I got acquainted with several libraries that I found on github, looked at how they work and how they are arranged. In fact, this is parsing and code collection.

It's probably doomed to fail, but I think it's worth a try. Negative experience = also experience.

I think I can do with script analysis (without saving additional data somewhere else). In addition, if (as I read here in the topics) we want to allow the user to edit the source code of the script, then he can simply erase the already important comments.

Now I took a simple script that created roundcube (that's all there is at the moment + examples from the documentation). So far it's working, but more checks and tests are needed. In any case, it's worth a try.

**Comment by plinss** [comment]
@sazanof given that sieve can be arbitrary script, unless your code can do human-level analysis, it's guaranteed to fail in many cases when reading script you haven't anticipated. This is going to lead to all sorts of failure modes and a bad user experience, especially considering that the vast majority of users don't (and shouldn't have to) understand how sieve even works.

There's also the situation where users use multiple different mail clients that all offer some kind of sieve script generation. Anything you generate will not be editable by other clients.

The best path forward here, is to work with the authors of other sieve generators and agree on a standard format of storing the metadata (or, at the least, output the format of the most popular client, and be able to understand the formats of the others). This way users can get a consistent and interoperable experience when moving between clients.

If you see a sieve script without metadata (or with metadata you can't parse), you should simply refuse to modify it. Possibly giving advanced users the option to load the script in a text editor (this should *not* be the default experience).

**Comment by loxK** [comment]
Please [see that comment](https://github.com/nextcloud/mail/issues/44#issuecomment-249711205).

**Comment by sazanof** [comment]
Hi! Guys, that's right. 
Not all libraries are written, not all libraries can be used. 

I completely agree that sieve is like a programmer's code - it's always different =). 

However, there is a specification. And if there is a specification, then there are rules. For example https://datatracker.ietf.org/doc/html/rfc5228 this specification.
You've probably met her. It is clear that the form of the language is abstract, but nevertheless there are tokens, comparators, conditions and rule blocks.

Of course, there are third-party extensions for sieve (spamtest and virus test). But nevertheless, there is a great opportunity to make basic support and a sieve rule configurator.

We could discuss this in more detail, because the truth is born in disputes. And in the discussions - the solution.

Please tell me, maybe someone has sieve files generated by other mail clients? Or files created and edited in several email clients at once? If there is, could you share them?

**Comment by dumblob** [comment]
@sazanof https://github.com/thsmi/sieve is the current state-of-the-art IMHO

**Comment by dumblob** [comment]
@plinss sure *thsmi/sieve* is quite low-level. Just two observations:

a) SIEVE standard itself is **deliberately** made flexible and **any** user has to accept that and realize that flexibility comes with a price which in this case is "use your brain"

b) what you are talking about is simply an analysis of the 5 most common use cases people have and then implementing a separate input form for each of these use cases - e.g. "autoreply to all senders during my vacation"

**Comment by plinss** [comment]
@dumblob 

> a) SIEVE standard itself is deliberately made flexible and any user has to accept that and realize that flexibility comes with a price which in this case is "use your brain"

I’m well aware of how sieve works, I’ve been a professional software engineer for over 40 years and I’m reasonably certain that I designed the code you’re using to read this (Google me if you doubt).

Saying “use your brain and learn to code” isn’t exactly user friendly.

> b) what you are talking about is simply an analysis of the 5 most common use cases people have and then implementing a separate input form for each of these use cases 

Yes, that’s *exactly* my point. Power users don’t need this feature at all, they can already figure out how to access and write sieve. This feature needs to be for the other 99% of the users *most common needs*. And it needs to be dead simple and obvious how to use.

**Comment by sazanof** [comment]
@plinss 
And hello again!
Thanks for this debate =)
> it'll be a great experiment and learning experience for you I'm sure

Undoubtedly =)

> but you don't update the metadata.

Question (I want to understand for myself). What do you mean by "metadata"? Could you explain in more detail. In my understanding, there is a text that needs to be analyzed, put together in a form convenient for the programmer, transfer it to the frontend, make a completed form out of it. Then the form can be changed, deleted or added or edited fields, conditions, etc.

After that, send the modified array to the server, where it is collected into the script text and later updated on the server.

Perhaps my experience is not enough to speak for all email clients, but I can judge some cloud clients and a couple of clients with a web interface.

Cloud clients (MS O365, Yandex mail), it seems to me, do not provide an opportunity from third-party IP to edit sieve files when connecting to an account.  Perhaps because the ports are open only on localhost (rough example)

By the way, I wanted to ask if you have considered using a separate sieve for NC? For example `~/nc-rules.sieve` for fillters and `~/nc-vacation.sieve` only for autoresponders. After all, NC is (a cloud platform on the user's side). These two scripts can be edited through the GUI, and the others (scripts created by Rainloop (it seems like it has its own scripts) and managesieve and others) can be edited through the code editor or not output at all. 
Yes, these are potential problems and confusion on the one hand (I wrote one script in one GUI, and there is no script in the second (example)), but on the other hand, at least this way it is possible to separate, secure the writing and processing of scripts inside NC. Well, if a third email client appears tomorrow with the ability to edit sieve scripts with their hands, then "advanced" users will certainly be able to break it when editing in textarea (I did it =)). Then I searched for a long time where my mail was in 2 days. Found.

Well, I had to ask, understand.

Thank you and have a nice day to everyone!

**Comment by ChristophWurst** [comment]
> By the way, I wanted to ask if you have considered using a separate sieve for NC? For example `~/nc-rules.sieve` for fillters and `~/nc-vacation.sieve` only for autoresponders. After all, NC is (a cloud platform on the user's side). These two scripts can be edited through the GUI, and the others (scripts created by Rainloop (it seems like it has its own scripts) and managesieve and others) can be edited through the code editor or not output at all.

IIRC there is always just one active script. E.g. if we added a script for the vacation auto repsonder and set is as active script, all other scripts get disabled.

**Comment by Luke1982** [comment]
Totally agree with @plinss. Using sieve in a 'regular non-technical user way' is a very common requirement for any mailclient these days. People don't even understand why "they can't do it from outlook".

**Comment by loxK** [comment]
> there is always just one active script.

True, and most webmails (like Roundcube or Rainloop) will use their own file and activate it. Which is annoying if you have been adding rules to another file: they won't be enabled anymore.

What every mail app should do is maintain a master sieve script that [includes the others](https://www.rfc-editor.org/rfc/rfc6609). Thus, activating a script file would rather mean being included by the master script.

master :

```
require "include";
include :personal "nexcloud";
include :personal "rainloop.user";
include :personal "roundcube";
```

**Comment by Bugsbane** [comment]
I see that this is a popular request, but one that hasn't seen even any discussion in almost a year. Have discussions on this been moved elsewhere? Is the bounty still active?

Personally I'm just looking for something like Thunderbird's Tools> Message filters feature (would also be awesome to have their show messages by starred / tagged filter, but that's just a bonus) . Currently I just fire up TB rergularly to auto filter my old stuff, but it would be great if Nextcloud Mail could take care of this.

I'm no coder, but happy to help out with any testing / mockups / UI stuff.

**Comment by loxK** [comment]
> Personally I'm just looking for something like Thunderbird's Tools> Message filters feature. Currently I just fire up TB rergularly to auto filter my old stuff, but it would be great if Nextcloud Mail could take care of this.

It is not the same, the sieve filtering happens on the server not the client, so that it doesn't depend on e-mails clients.

**Comment by BiglifeMatt** [comment]
Is the idea of an intuitive gui for sieve for typical end users still being considered here?

**Comment by loxK** [comment]
> Is the idea of an intuitive gui for sieve for typical end users still being considered here?

Sure, it is waiting for a contributor to build it.

**Comment by makeasnek** [comment]
Please do not use bountysource. Many devs have had trouble getting paid there. You can check out this lemmy community as an alternative https://lemmy.ml/c/bugbounties

For statements from devs who have been unable to cash out from bountysource see:
https://github.com/bountysource/core/issues

**Comment by PanderMusubi** [comment]
Nevertheless, I hope somebody claims it and the money doesn't go to Bountysource.

**Comment by gabviv73** [comment]
IT would be nice to configure the UI for standard filters 'à là'  Horde in Ingo. 
NC Mail already have Autoresponder,  could be added:
-  Forward
- Whitelist
- Blacklist
See Horde screenshot:
![image](https://github.com/nextcloud/mail/assets/1001288/5cc4dc34-580d-4f58-aa99-ea8b2116b9f7)

**Comment by kesselb** [comment]
Hello everyone, we’ll be releasing the Sieve interface with the upcoming Mail 4.1 update soon! If you encounter any issues or notice anything missing, please open a new issue. When reporting an issue, kindly include the generated Sieve script and metadata if possible to help us resolve it. Thank you!
